File: docs/conduit_engine.js
// CONDUIT — prototype engine (JS port of conduit_engine.py). Fiber flow game on a degree-6 hex-cell
// board. Real Dinic max-flow + min-cut drive scoring, the AI, and the UI's "danger line". Offline.
// Cells are keyed "q,r"; edges keyed "a|b" (endpoints sorted). Two players: 0 = you, 1 = AI.
(function (root, factory) {
  var mod = factory();
  if (typeof module !== "undefined" && module.exports) module.exports = mod;
  root.CONDUIT = mod;
})(typeof self !== "undefined" ? self : this, function () {
  var DIRS = [[1, 0], [1, -1], [0, -1], [-1, 0], [-1, 1], [0, 1]];
  var COST = { build: 3, upgrade: 2, interdict: 3 };
  var MAXCAP = 3, TICKS = 20;

  function ck(q, r) { return q + "," + r; }
  function edge(a, b) { return a < b ? a + "|" + b : b + "|" + a; }
  function ends(e) { return e.split("|"); }

  function genBoard(R) {
    var cells = [], set = {};
    for (var q = -R; q <= R; q++) for (var r = -R; r <= R; r++)
      if (Math.abs(q + r) <= R) { var k = ck(q, r); cells.push(k); set[k] = true; }
    var adj = {};
    cells.forEach(function (k) {
      var p = k.split(",").map(Number), out = [];
      DIRS.forEach(function (d) { var nk = ck(p[0] + d[0], p[1] + d[1]); if (set[nk]) out.push(nk); });
      adj[k] = out;
    });
    return { cells: cells, set: set, adj: adj };
  }

  // symmetric fiber scenario: opposite-corner PoPs, exchanges in 180deg pairs + a fat centre
  function makeScenario(R) {
    R = R || 3;
    var b = genBoard(R);
    var srcA = ck(-R, R), srcB = ck(R, -R);   // player 0 (you) in the BOTTOM-LEFT, matching the other games
    var raw = [[[0, 0], 4, 3], [[R - 1, -1], 3, 3], [[1, R - 2], 2, 4]];
    var sinks = [];
    raw.forEach(function (s, i) {
      var c = s[0];
      function push(cell) { var k = ck(cell[0], cell[1]); if (b.set[k]) sinks.push({ cell: k, dem: s[1], price: s[2] }); }
      push(c);
      if (i > 0) push([-c[0], -c[1]]);   // its 180deg mirror
    });
    return { cells: b.cells, adj: b.adj, set: b.set, src: [srcA, srcB], sinks: sinks, R: R };
  }

  // ---------------- Dinic max-flow ----------------
  function Dinic(n) { this.n = n; this.g = []; for (var i = 0; i < n; i++) this.g.push([]); }
  Dinic.prototype.add = function (u, v, c) {
    this.g[u].push([v, c, this.g[v].length]);
    this.g[v].push([u, 0, this.g[u].length - 1]);
  };
  Dinic.prototype.bfs = function (s, t) {
    this.lv = new Int32Array(this.n).fill(-1); this.lv[s] = 0; var q = [s], h = 0;
    while (h < q.length) { var u = q[h++]; for (var i = 0; i < this.g[u].length; i++) { var e = this.g[u][i]; if (e[1] > 0 && this.lv[e[0]] < 0) { this.lv[e[0]] = this.lv[u] + 1; q.push(e[0]); } } }
    return this.lv[t] >= 0;
  };
  Dinic.prototype.dfs = function (u, t, f) {
    if (u === t) return f;
    for (; this.it[u] < this.g[u].length; this.it[u]++) {
      var e = this.g[u][this.it[u]];
      if (e[1] > 0 && this.lv[e[0]] === this.lv[u] + 1) {
        var d = this.dfs(e[0], t, Math.min(f, e[1]));
        if (d > 0) { e[1] -= d; this.g[e[0]][e[2]][1] += d; return d; }
      }
    }
    return 0;
  };
  Dinic.prototype.maxflow = function (s, t) {
    var flow = 0;
    while (this.bfs(s, t)) { this.it = new Int32Array(this.n); var f; while ((f = this.dfs(s, t, 1e9)) > 0) flow += f; }
    return flow;
  };
  Dinic.prototype.reachable = function (s) {   // residual-reachable from s => source side of a min-cut
    var seen = new Uint8Array(this.n); seen[s] = 1; var q = [s], h = 0;
    while (h < q.length) { var u = q[h++]; for (var i = 0; i < this.g[u].length; i++) { var e = this.g[u][i]; if (e[1] > 0 && !seen[e[0]]) { seen[e[0]] = 1; q.push(e[0]); } } }
    return seen;
  };

  function idxMap(cells) { var m = {}; for (var i = 0; i < cells.length; i++) m[cells[i]] = i; return m; }

  function buildGraph(links, cells, extra) {
    var idx = idxMap(cells), d = new Dinic(cells.length + (extra || 0));
    for (var e in links) { var c = links[e]; if (c <= 0) continue; var uv = ends(e); d.add(idx[uv[0]], idx[uv[1]], c); d.add(idx[uv[1]], idx[uv[0]], c); }
    return { d: d, idx: idx };
  }

  function flowToOne(links, source, sink, cells) {
    if (source === sink) return 0;
    var g = buildGraph(links, cells, 0);
    return g.d.maxflow(g.idx[source], g.idx[sink]);
  }

  // max-flow to a super-sink, each exchange feeding it at its contended share; returns {thr, cut:[edges]}
  function delivered(links, source, sinkCaps, cells) {
    var g = buildGraph(links, cells, 1), T = cells.length, any = false;
    for (var s in sinkCaps) { if (sinkCaps[s] > 0) { g.d.add(g.idx[s], T, sinkCaps[s]); any = true; } }
    if (!(source in g.idx) || !any) return { thr: 0, cut: [] };
    var thr = g.d.maxflow(g.idx[source], T);
    // min-cut: edges from residual-reachable set to the rest that are real (non-super-sink) links
    var seen = g.d.reachable(g.idx[source]), cut = [], inv = cells;
    for (var e in links) { var uv = ends(e); var a = g.idx[uv[0]], b = g.idx[uv[1]]; if (links[e] > 0 && seen[a] !== seen[b]) cut.push(e); }
    return { thr: thr, cut: cut };
  }

  // network capacity: how much p COULD deliver if they won every exchange's full demand. Capped by both
  // the board's demand and p's own min-cut — so it's the right "robustness" target for the AI to invest
  // surplus credit into (a fatter, cut-resilient net) instead of hoarding, and it drives upgrades.
  function netCapacity(state, p) {
    var full = {}; state.sc.sinks.forEach(function (sk) { full[sk.cell] = sk.dem; });
    return delivered(effLinks(state, p), state.sc.src[p], full, state.sc.cells).thr;
  }

  // DANGER SET: every one of p's links where a single enemy Cut (a -1 throttle, exactly the in-game
  // interdiction) would actually reduce p's DELIVERED throughput — computed by applying that throttle and
  // re-resolving the full game (so contention shifts are included), then comparing. On a uniform 1-wide
  // path that's the WHOLE path (all equally critical — no arbitrary "first link"); once a parallel route
  // carries the load, a cut that costs nothing drops off the set. This is what the UI's red line shows.
  function dangerLinks(state, p) {
    var base = resolve(state).thr[p];
    if (base <= 0) return [];
    var out = [], L = state.links[p];
    for (var e in L) {
      var key = p + "@" + e, had = key in state.throttle;
      if (!had) state.throttle[key] = 3;                 // simulate the Cut (-1 capacity on p's link e)
      var t = resolve(state).thr[p];
      if (!had) delete state.throttle[key];
      if (t < base) out.push(e);
    }
    return out;
  }

  // effective links = a player's links with active interdiction throttles applied
  function effLinks(state, p) {
    var out = {}, L = state.links[p];
    for (var e in L) out[e] = L[e];
    for (var key in state.throttle) { var parts = key.split("@"); if (+parts[0] === p && out[parts[1]] != null) out[parts[1]] = Math.max(0, out[parts[1]] - 1); }
    return out;
  }

  // resolve one tick: contention split by reach, then delivered max-flow per player. Returns throughput,
  // per-exchange split, and each player's min-cut edge set (for the UI danger line).
  function resolve(state) {
    var sc = state.sc, cells = sc.cells, eff = [effLinks(state, 0), effLinks(state, 1)];
    var reach = [{}, {}], caps = [{}, {}], split = {};
    sc.sinks.forEach(function (sk) {
      var a = flowToOne(eff[0], sc.src[0], sk.cell, cells), b = flowToOne(eff[1], sc.src[1], sk.cell, cells);
      reach[0][sk.cell] = a; reach[1][sk.cell] = b;
      var tot = a + b, ca = 0, cb = 0;
      if (tot > 0) { ca = Math.floor(sk.dem * a / tot); cb = Math.floor(sk.dem * b / tot); var rem = sk.dem - ca - cb; if (rem) { if (a >= b) ca += rem; else cb += rem; } }
      caps[0][sk.cell] = ca; caps[1][sk.cell] = cb; split[sk.cell] = [ca, cb];
    });
    var d0 = delivered(eff[0], sc.src[0], caps[0], cells), d1 = delivered(eff[1], sc.src[1], caps[1], cells);
    return { thr: [d0.thr, d1.thr], split: split, cut: [d0.cut, d1.cut] };
  }

  // ---------------- game state ----------------
  function newGame(R) {
    var sc = makeScenario(R || 3);
    return { sc: sc, links: [{}, {}], credit: [12, 12], revenue: [0, 0], throttle: {}, tick: 0, ticks: TICKS };
  }

  function touch(state, p) { var s = {}; s[state.sc.src[p]] = 1; for (var e in state.links[p]) { var uv = ends(e); s[uv[0]] = 1; s[uv[1]] = 1; } return s; }

  function legalBuilds(state, p) {
    var t = touch(state, p), occ = {}, e; for (e in state.links[0]) occ[e] = 1; for (e in state.links[1]) occ[e] = 1;
    var out = {}, adj = state.sc.adj;
    for (var c in t) adj[c].forEach(function (nb) { var e2 = edge(c, nb); if (!occ[e2]) out[e2] = 1; });
    return Object.keys(out);
  }
  function legalUpgrades(state, p) { var out = []; for (var e in state.links[p]) if (state.links[p][e] < MAXCAP) out.push(e); return out; }
  function legalInterdicts(state, p) {
    var opp = 1 - p, t = touch(state, p), out = [];
    for (var e in state.links[opp]) { var uv = ends(e); if (state.links[opp][e] > 0 && (t[uv[0]] || t[uv[1]]) && !(opp + "@" + e in state.throttle)) out.push(e); }
    return out;
  }

  function canAfford(state, p, kind) { return state.credit[p] >= COST[kind]; }

  function apply(state, p, kind, e) {
    if (state.credit[p] < COST[kind]) return false;
    if (kind === "build") { if (state.links[0][e] || state.links[1][e]) return false; state.links[p][e] = 1; }
    else if (kind === "upgrade") { if (!state.links[p][e] || state.links[p][e] >= MAXCAP) return false; state.links[p][e] += 1; }
    else if (kind === "interdict") { state.throttle[(1 - p) + "@" + e] = 3; }
    else return false;
    state.credit[p] -= COST[kind];
    return true;
  }

  function tickIncome(state, thr) {
    for (var p = 0; p < 2; p++) { state.revenue[p] += thr[p]; state.credit[p] += 4 + thr[p]; }
    for (var k in state.throttle) { state.throttle[k] -= 1; if (state.throttle[k] <= 0) delete state.throttle[k]; }
  }

  // ---------------- max-flow AI (greedy 1-ply + path shaping) ----------------
  function reachPotential(state, p) {
    var t = touch(state, p), opp = state.links[1 - p], adj = state.sc.adj, dist = {}, q = [], h = 0;
    for (var c in t) { dist[c] = 0; q.push(c); }
    while (h < q.length) { var u = q[h++]; adj[u].forEach(function (nb) { if (opp[edge(u, nb)]) return; if (dist[nb] == null) { dist[nb] = dist[u] + 1; q.push(nb); } }); }
    var pot = 0; state.sc.sinks.forEach(function (sk) { pot -= (dist[sk.cell] != null ? dist[sk.cell] : 12); });
    return pot;
  }

  function snapshot(state) { return { l0: Object.assign({}, state.links[0]), l1: Object.assign({}, state.links[1]), thr: Object.assign({}, state.throttle), cr: state.credit.slice() }; }
  function restore(state, s) { state.links[0] = s.l0; state.links[1] = s.l1; state.throttle = s.thr; state.credit = s.cr; }

  function aiTurn(state, p) {
    var budget = 3, SHAPE = 0.2, ROBUST = 0.12;   // ROBUST: value network capacity so surplus credit buys
    while (budget > 0 && state.credit[p] >= 2) {   // a fatter, cut-resilient net (drives upgrades) not hoarding
      var base = resolve(state), b0 = (base.thr[p] - base.thr[1 - p]) + SHAPE * reachPotential(state, p) + ROBUST * netCapacity(state, p);
      var cands = [];
      legalBuilds(state, p).forEach(function (e) { if (canAfford(state, p, "build")) cands.push(["build", e]); });
      legalUpgrades(state, p).forEach(function (e) { if (canAfford(state, p, "upgrade")) cands.push(["upgrade", e]); });
      legalInterdicts(state, p).forEach(function (e) { if (canAfford(state, p, "interdict")) cands.push(["interdict", e]); });
      if (!cands.length) break;
      // shuffle + cap candidate count so the JS turn stays snappy
      for (var i = cands.length - 1; i > 0; i--) { var j = (Math.random() * (i + 1)) | 0; var tmp = cands[i]; cands[i] = cands[j]; cands[j] = tmp; }
      if (cands.length > 44) cands = cands.slice(0, 44);
      var best = null, bestGain = 1e-9;
      for (var ci = 0; ci < cands.length; ci++) {
        var mv = cands[ci], s = snapshot(state);
        if (!apply(state, p, mv[0], mv[1])) { restore(state, s); continue; }
        var t = resolve(state), gain = (t.thr[p] - t.thr[1 - p]) + SHAPE * reachPotential(state, p) + ROBUST * netCapacity(state, p) - b0;
        if (mv[0] === "interdict") gain -= 0.25;
        restore(state, s);
        if (gain > bestGain) { bestGain = gain; best = mv; }
      }
      if (!best) {
        // greedy plateaued (widening a cut needs several upgrades, each flat on its own). If flush with
        // credit, invest it in robustness: upgrade a bottleneck (danger) link, else a flow-carrying link.
        if (state.credit[p] >= COST.upgrade + 2) {
          var dl = dangerLinks(state, p), up = null;
          for (var di = 0; di < dl.length; di++) if (state.links[p][dl[di]] < MAXCAP) { up = dl[di]; break; }
          if (!up) { var ups = legalUpgrades(state, p); if (ups.length) up = ups[(Math.random() * ups.length) | 0]; }
          if (up) { apply(state, p, "upgrade", up); budget--; continue; }
        }
        break;
      }
      apply(state, p, best[0], best[1]); budget--;
    }
  }

  return {
    makeScenario: makeScenario, newGame: newGame, resolve: resolve,
    legalBuilds: legalBuilds, legalUpgrades: legalUpgrades, legalInterdicts: legalInterdicts,
    canAfford: canAfford, apply: apply, tickIncome: tickIncome, aiTurn: aiTurn, dangerLinks: dangerLinks,
    ends: ends, edge: edge, COST: COST, MAXCAP: MAXCAP, TICKS: TICKS
  };
});

File: conduit_engine.py
"""CONDUIT — prototype engine + max-flow evaluator + self-play balance sim.

A fiber-backhaul flow game on a DEGREE-6 hex-cell board. Each player builds their own capacitated link
network from a corner PoP to shared exchanges (sinks); each tick the game computes the MAX-FLOW each
player can deliver (capped by their own min-cut) and pays for throughput. See CONDUIT.md for the design.

The point of this file: prove the core is (a) computable, (b) non-degenerate, (c) roughly fair, using a
principled max-flow AI in self-play. Run:  python3 conduit_engine.py [nGames] [radius] [ticks]
"""
import sys, random, collections

A, B = 0, 1                      # players
DIRS = [(1, 0), (1, -1), (0, -1), (-1, 0), (-1, 1), (0, 1)]   # hex axial neighbours (degree 6)


# ----------------------------------------------------------------------------- board
def gen_board(R):
    cells = [(q, r) for q in range(-R, R + 1) for r in range(-R, R + 1) if abs(q + r) <= R]
    cellset = set(cells)
    adj = {c: [(c[0] + d[0], c[1] + d[1]) for d in DIRS if (c[0] + d[0], c[1] + d[1]) in cellset] for c in cells}
    return cells, cellset, adj


def make_scenario(R):
    """Symmetric fiber scenario: two opposite-corner PoPs, exchanges placed in 180deg-symmetric pairs
    plus a high-demand centre. Negating axial coords is a 180deg rotation, so the map is fair."""
    cells, cellset, adj = gen_board(R)
    srcA = (-R, R); srcB = (R, -R)                      # player 0 (you) bottom-left, matching the other games
    # exchanges: (cell, demand, price). centre is the fat contested one; two symmetric mid pairs.
    raw = [((0, 0), 4, 3)]
    for cell, dem, price in [((R - 1, -1), 3, 3), ((1, R - 2), 2, 4)]:
        raw.append((cell, dem, price))
        raw.append(((-cell[0], -cell[1]), dem, price))   # its 180deg mirror
    sinks = [(c, d, p) for (c, d, p) in raw if c in cellset]
    return dict(cells=cells, cellset=cellset, adj=adj, src=[srcA, srcB], sinks=sinks)


# ----------------------------------------------------------------------------- Dinic max-flow
class Dinic:
    def __init__(self, n):
        self.n = n; self.g = [[] for _ in range(n)]
    def add(self, u, v, c):                              # directed arc u->v cap c (+ 0-cap reverse)
        self.g[u].append([v, c, len(self.g[v])])
        self.g[v].append([u, 0, len(self.g[u]) - 1])
    def _bfs(self, s, t):
        self.lv = [-1] * self.n; self.lv[s] = 0; q = collections.deque([s])
        while q:
            u = q.popleft()
            for v, c, _ in self.g[u]:
                if c > 0 and self.lv[v] < 0:
                    self.lv[v] = self.lv[u] + 1; q.append(v)
        return self.lv[t] >= 0
    def _dfs(self, u, t, f):
        if u == t: return f
        while self.it[u] < len(self.g[u]):
            e = self.g[u][self.it[u]]
            v, c, rev = e
            if c > 0 and self.lv[v] == self.lv[u] + 1:
                d = self._dfs(v, t, min(f, c))
                if d > 0:
                    e[1] -= d; self.g[v][rev][1] += d; return d
            self.it[u] += 1
        return 0
    def maxflow(self, s, t):
        flow = 0
        while self._bfs(s, t):
            self.it = [0] * self.n
            while True:
                f = self._dfs(s, t, float("inf"))
                if not f: break
                flow += f
        return flow
    def min_cut_reachable(self, s):                     # residual-reachable set from s => source side of a min-cut
        seen = [False] * self.n; seen[s] = True; q = collections.deque([s])
        while q:
            u = q.popleft()
            for v, c, _ in self.g[u]:
                if c > 0 and not seen[v]: seen[v] = True; q.append(v)
        return seen


def _build_graph(links, cells, extra_nodes=0):
    idx = {c: i for i, c in enumerate(cells)}
    d = Dinic(len(cells) + extra_nodes)
    for (u, v), cap in links.items():                   # undirected fiber => an arc each way at full cap
        d.add(idx[u], idx[v], cap); d.add(idx[v], idx[u], cap)
    return d, idx


def flow_to_one(links, source, sink, cells):
    """Max reach from `source` to a single `sink` in this player's network (uncapped sink)."""
    if source == sink: return 0
    d, idx = _build_graph(links, cells)
    return d.maxflow(idx[source], idx[sink])


def delivered(links, source, sink_caps, cells):
    """Max-flow from source to a super-sink, each exchange feeding it at its contended share.
    Returns (throughput, dinic, idx, T) so callers can pull the min-cut if they want."""
    d, idx = _build_graph(links, cells, extra_nodes=1)
    T = len(cells)                                      # super-sink index
    any_cap = False
    for s, cap in sink_caps.items():
        if cap > 0: d.add(idx[s], T, cap); any_cap = True
    if source not in idx or not any_cap: return 0, d, idx, T
    return d.maxflow(idx[source], T), d, idx, T


# ----------------------------------------------------------------------------- game state + flow resolve
class Game:
    def __init__(self, sc, ticks=24):
        self.sc = sc; self.ticks = ticks
        self.links = [dict(), dict()]                   # per player: {(u,v) sorted : capacity}
        self.credit = [12, 12]                          # equal start; measure the natural move-order edge first
        self.revenue = [0, 0]                           # cumulative score
        self.throttle = {}                              # (player,(u,v)) -> ticks remaining
        self.tick = 0

    # ---- flow resolution: contended shares, then each player's delivered max-flow ----
    def resolve(self):
        sc = self.sc
        eff = [self._effective_links(p) for p in (A, B)]
        reach = [{}, {}]
        for (cell, dem, price) in sc["sinks"]:
            for p in (A, B):
                reach[p][cell] = flow_to_one(eff[p], sc["src"][p], cell, sc["cells"])
        capA, capB = {}, {}
        for (cell, dem, price) in sc["sinks"]:
            a, b = reach[A][cell], reach[B][cell]; tot = a + b
            if tot == 0: capA[cell] = capB[cell] = 0; continue
            ca = dem * a // tot; cb = dem * b // tot
            rem = dem - ca - cb                          # give the leftover unit to the stronger reach
            if rem:
                if a >= b: ca += rem
                else: cb += rem
            capA[cell], capB[cell] = ca, cb
        thr = [0, 0]
        for p, caps in ((A, capA), (B, capB)):
            t, *_ = delivered(eff[p], sc["src"][p], caps, sc["cells"])
            thr[p] = t
        return thr, (capA, capB)

    def _effective_links(self, p):
        """This player's links with active throttles applied (interdiction lowers capacity)."""
        out = dict(self.links[p])
        for (pl, e), _ in self.throttle.items():
            if pl == p and e in out: out[e] = max(0, out[e] - 1)
        return out

    # ---- helpers ----
    def _touch(self, p):                                # cells the player's network reaches (for adjacency rules)
        s = {self.sc["src"][p]}
        for (u, v) in self.links[p]: s.add(u); s.add(v)
        return s

    def legal_builds(self, p):
        touched = self._touch(p); occupied = set(self.links[A]) | set(self.links[B])
        out = []
        for c in touched:
            for nb in self.sc["adj"][c]:
                e = tuple(sorted((c, nb)))
                if e not in occupied: out.append(e)
        return list(set(out))

    def legal_upgrades(self, p):
        return [e for e, cap in self.links[p].items() if cap < 3]

    def legal_interdicts(self, p):
        opp = 1 - p; touched = self._touch(p); out = []
        for e, cap in self.links[opp].items():
            if cap > 0 and (e[0] in touched or e[1] in touched) and (opp, e) not in self.throttle:
                out.append(e)
        return out

    COST = dict(build=3, upgrade=2, interdict=3)

    def apply(self, p, mv):
        kind, e = mv
        c = self.COST[kind]
        if self.credit[p] < c: return False
        if kind == "build":
            if e in self.links[A] or e in self.links[B]: return False
            self.links[p][e] = 1
        elif kind == "upgrade":
            if e not in self.links[p] or self.links[p][e] >= 3: return False
            self.links[p][e] += 1
        elif kind == "interdict":
            self.throttle[(1 - p, e)] = 3               # heals after 3 ticks
        self.credit[p] -= c
        return True

    def clone(self):
        g = Game.__new__(Game); g.sc = self.sc; g.ticks = self.ticks
        g.links = [dict(self.links[0]), dict(self.links[1])]; g.credit = list(self.credit)
        g.revenue = list(self.revenue); g.throttle = dict(self.throttle); g.tick = self.tick
        return g

    def tick_income(self, thr):
        for p in (A, B):
            self.revenue[p] += thr[p]                    # cumulative score = total delivered bandwidth
            self.credit[p] += 4 + thr[p]                 # base + revenue reinvested
        # decay throttles
        for k in list(self.throttle):
            self.throttle[k] -= 1
            if self.throttle[k] <= 0: del self.throttle[k]


# ----------------------------------------------------------------------------- max-flow AI (greedy 1-ply)
def reach_potential(g, p):
    """Shaping term: -Σ hop-distance from the player's network to each exchange, through edges not
    blocked by the opponent. Gives the greedy AI a gradient to LAY PATHS before any flow exists
    (a single link delivers 0 throughput, so without this the AI never starts)."""
    touched = g._touch(p); opp_edges = set(g.links[1 - p])
    dist = {c: 0 for c in touched}; q = collections.deque(touched)
    while q:
        u = q.popleft()
        for nb in g.sc["adj"][u]:
            if tuple(sorted((u, nb))) in opp_edges: continue     # can't build/route through enemy fiber
            if nb not in dist: dist[nb] = dist[u] + 1; q.append(nb)
    return -sum(dist.get(cell, 12) for (cell, dem, price) in g.sc["sinks"])


def ai_take_turn(g, p, rng):
    """Spend the action budget greedily: each action, pick the legal move maximising
    (myThroughput - oppThroughput) + a small path-progress shaping term, both from a real max-flow
    resolve. Throughput dominates once connected; shaping only breaks the zero-gradient early."""
    budget = 3; SHAPE = 0.2
    while budget > 0 and g.credit[p] >= min(g.COST.values()):
        base_thr, _ = g.resolve()
        base = (base_thr[p] - base_thr[1 - p]) + SHAPE * reach_potential(g, p)
        cands = ([("build", e) for e in g.legal_builds(p) if g.credit[p] >= g.COST["build"]] +
                 [("upgrade", e) for e in g.legal_upgrades(p) if g.credit[p] >= g.COST["upgrade"]] +
                 [("interdict", e) for e in g.legal_interdicts(p) if g.credit[p] >= g.COST["interdict"]])
        if not cands: break
        rng.shuffle(cands)                               # break ties randomly for variety
        best, best_gain = None, 1e-9                     # require a strictly positive marginal gain
        for mv in cands:
            snap = [dict(g.links[0]), dict(g.links[1])]; snap_thr = dict(g.throttle); snap_cred = list(g.credit)
            if not g.apply(p, mv):
                g.links = snap; g.throttle = snap_thr; g.credit = snap_cred; continue
            t, _ = g.resolve()
            gain = (t[p] - t[1 - p]) + SHAPE * reach_potential(g, p) - base
            if mv[0] == "interdict": gain -= 0.25        # an interdict must clearly beat building
            g.links = snap; g.throttle = snap_thr; g.credit = snap_cred      # restore
            if gain > best_gain: best_gain, best = gain, mv
        if best is None: break
        g.apply(p, best); budget -= 1


# ----------------------------------------------------------------------------- self-play balance sim
def play_game(sc, ticks, rng, swap=False, first_bonus=0, simultaneous=False):
    g = Game(sc, ticks)
    first = B if swap else A                             # who moves first this game
    g.credit[first] += first_bonus                       # compensation knob for the reaction advantage
    order = [first, 1 - first]
    thr_hist = []
    for _ in range(ticks):
        if simultaneous:
            # both decide against the SAME pre-tick state (no reaction), then apply together;
            # `first` wins any edge both tried to build.
            ga = g.clone(); ai_take_turn(ga, order[0], rng)
            gb = g.clone(); ai_take_turn(gb, order[1], rng)
            p0, p1 = order
            l0, l1 = ga.links[p0], gb.links[p1]
            for e in set(l0) & set(l1):                  # edge conflict -> p0 (the "first" decider) keeps it
                if e not in g.links[p0] and e not in g.links[p1]: del l1[e]
            g.links[p0] = l0; g.links[p1] = l1
            g.credit[p0] = ga.credit[p0]; g.credit[p1] = gb.credit[p1]
            g.throttle = {**ga.throttle, **gb.throttle}
        else:
            for p in order:
                ai_take_turn(g, p, rng)
        thr, _ = g.resolve()
        g.tick_income(thr); g.tick += 1
        thr_hist.append(tuple(thr))
    return g, thr_hist


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 30
    R = int(sys.argv[2]) if len(sys.argv) > 2 else 3
    ticks = int(sys.argv[3]) if len(sys.argv) > 3 else 20
    sc = make_scenario(R)
    print(f"CONDUIT sim: R={R} ({len(sc['cells'])} cells, deg-6), {len(sc['sinks'])} exchanges, "
          f"{ticks} ticks, {n} games")
    fb = int(sys.argv[4]) if len(sys.argv) > 4 else 0   # first-mover credit compensation (balance knob)
    rng = random.Random(7)
    firstwins = 0; margins = []; finals = []; mincuts = []
    for i in range(n):
        swap = (i % 2 == 1)                              # alternate who is "first" to isolate first-mover edge
        sim = "sim" in sys.argv                          # pass 'sim' to test simultaneous-turn resolution
        g, hist = play_game(sc, ticks, rng, swap=swap, first_bonus=fb, simultaneous=sim)
        first = B if swap else A
        rf, rs = g.revenue[first], g.revenue[1 - first]
        if rf > rs: firstwins += 1
        margins.append(abs(g.revenue[0] - g.revenue[1]))
        finals.append(tuple(g.revenue))
        # measure a network's min-cut width at game end (how fragile are the built nets?)
        for p in (A, B):
            _, cap = g.resolve()
            t, d, idx, T = delivered(g._effective_links(p), sc["src"][p], cap[p], sc["cells"])
            mincuts.append(t)
    import statistics as st
    print(f"  first-mover win-rate : {firstwins}/{n} = {firstwins/n:.2f}   (0.5 = fair)")
    print(f"  avg victory margin   : {st.mean(margins):.1f} bandwidth-units cumulative")
    print(f"  avg final revenue    : {st.mean(a for a,_ in finals):.0f} vs {st.mean(b for _,b in finals):.0f}")
    print(f"  delivered/tick spread: min {min(mincuts)}  median {int(st.median(mincuts))}  max {max(mincuts)}"
          f"   (want a spread, not all pinned at 1 or all huge)")
    # show one sample game's throughput trajectory
    g, hist = play_game(sc, ticks, rng)
    print("  sample throughput/tick (A,B):", " ".join(f"{a},{b}" for a, b in hist))


if __name__ == "__main__":
    main()
